From 229bdc116fe22d76a3611eb242a3493c4d020df6 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 17 Sep 2023 16:53:15 +0200
Subject: [PATCH 0106/1010] genirq: irq_pipeline: introduce
 smp_call_function_oob()

This new call is logically equivalent to smp_call_function_single(),
except that function handlers run on the out-of-band stage.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/linux/smp.h      |  20 ++++++
 kernel/irq/irqptorture.c |  95 ++++++++++++++++++++++----
 kernel/irq/pipeline.c    |  58 ++++++++++++++--
 kernel/smp.c             | 142 +++++++++++++++++++++++++++++++++++++++
 4 files changed, 298 insertions(+), 17 deletions(-)

diff --git a/include/linux/smp.h b/include/linux/smp.h
index 12be3aca4c..4474996b03 100644
--- a/include/linux/smp.h
+++ b/include/linux/smp.h
@@ -275,6 +275,20 @@ static inline int get_boot_cpu_id(void)
 		raw_smp_processor_id();			\
 	})
 #define hard_put_cpu(flags)	hard_preempt_enable(flags)
+
+int up_oob_call(smp_call_func_t func, void *info);
+
+#ifdef CONFIG_SMP
+int smp_call_function_oob(int cpu, smp_call_func_t func, void *info, bool wait);
+void smp_flush_oob_call_function_queue(void);
+#else
+static inline int smp_call_function_oob(int cpu, smp_call_func_t func,
+					void *info, bool wait)
+{
+	return up_oob_call(func, info);
+}
+#endif /* !CONFIG_SMP */
+
 #else
 #define hard_get_cpu(flags)	({ (void)(flags); get_cpu(); })
 #define hard_put_cpu(flags)	\
@@ -282,6 +296,12 @@ static inline int get_boot_cpu_id(void)
 		(void)(flags);	\
 		put_cpu();	\
 	} while (0)
+
+static inline int smp_call_function_oob(int cpu, smp_call_func_t func,
+					void *info, bool wait)
+{
+	return -ENOTSUPP;
+}
 #endif
 
 /*
diff --git a/kernel/irq/irqptorture.c b/kernel/irq/irqptorture.c
index f4ad93d07a..ddd4a9c2bd 100644
--- a/kernel/irq/irqptorture.c
+++ b/kernel/irq/irqptorture.c
@@ -3,22 +3,22 @@
  *
  * Copyright (C) 2017 Philippe Gerum  <rpm@xenomai.org>.
  */
-#include <linux/kernel.h>
-#include <linux/ktime.h>
-#include <linux/torture.h>
-#include <linux/printk.h>
-#include <linux/delay.h>
-#include <linux/tick.h>
-#include <linux/smp.h>
-#include <linux/cpumask.h>
 #include <linux/clockchips.h>
+#include <linux/completion.h>
+#include <linux/cpumask.h>
+#include <linux/delay.h>
 #include <linux/interrupt.h>
 #include <linux/irq.h>
 #include <linux/irq_pipeline.h>
-#include <linux/stop_machine.h>
 #include <linux/irq_work.h>
-#include <linux/completion.h>
+#include <linux/kernel.h>
+#include <linux/ktime.h>
+#include <linux/printk.h>
 #include <linux/slab.h>
+#include <linux/smp.h>
+#include <linux/stop_machine.h>
+#include <linux/tick.h>
+#include <linux/torture.h>
 #include "settings.h"
 
 static void torture_event_handler(struct clock_event_device *dev)
@@ -70,8 +70,8 @@ static int stop_machine_handler(void *arg)
 }
 
 /*
- * We test stop_machine() as a way to validate IPI handling in a
- * pipelined interrupt context.
+ * We test stop_machine() as a way to validate IPI broadcast when irqs
+ * are pipelined.
  */
 static int test_stop_machine(void)
 {
@@ -120,6 +120,73 @@ static int test_stop_machine(void)
 	return ret;
 }
 
+struct call_function_p_data {
+	int origin_cpu;
+	cpumask_var_t oob_mask;
+};
+
+static void remote_function_handler(void *arg)
+{
+	struct call_function_p_data *p = arg;
+	int cpu = raw_smp_processor_id();
+
+	if (running_oob())
+		cpumask_set_cpu(cpu, p->oob_mask);
+
+	if (cpu != p->origin_cpu)
+		pr_alert("irq_pipeline" TORTURE_FLAG
+			 " CPU%d handles remote function call\n", cpu);
+}
+
+static int test_call_function_oob(void)
+{
+	struct call_function_p_data d;
+	cpumask_var_t tmp_mask;
+	int ret = -ENOMEM, cpu;
+
+	if (!zalloc_cpumask_var(&d.oob_mask, GFP_KERNEL)) {
+		WARN_ON(1);
+		return ret;
+	}
+
+	if (!alloc_cpumask_var(&tmp_mask, GFP_KERNEL)) {
+		WARN_ON(1);
+		goto fail;
+	}
+
+	ret = -EINVAL;
+	d.origin_cpu = raw_smp_processor_id();
+	pr_alert("irq_pipeline" TORTURE_FLAG
+		 " CPU%d calls function on remote(s)\n",
+		 d.origin_cpu);
+
+	for_each_cpu(cpu, cpu_online_mask) {
+		/* Issue a synchronous call to each online CPU. */
+		ret = smp_call_function_oob(cpu, remote_function_handler, &d, true);
+		WARN_ON(ret);
+		if (ret)
+			goto fail;
+	}
+
+	/*
+	 * Check whether all callback functions did run as expected on
+	 * the oob stage.
+	 */
+	cpumask_xor(tmp_mask, cpu_online_mask, d.oob_mask);
+	if (!cpumask_empty(tmp_mask)) {
+		for_each_cpu(cpu, tmp_mask)
+			pr_alert("irq_pipeline" TORTURE_FLAG
+				" CPU%d: remote function call did not run out-of-band!\n",
+				cpu);
+	}
+
+	free_cpumask_var(tmp_mask);
+fail:
+	free_cpumask_var(d.oob_mask);
+
+	return ret;
+}
+
 static struct irq_work_tester {
 	struct irq_work work;
 	struct completion done;
@@ -237,6 +304,10 @@ static int __init irqp_torture_init(void)
 	if (ret)
 		goto out;
 
+	ret = test_call_function_oob();
+	if (ret)
+		goto out;
+
 	ret = start_tick_takeover_test();
 	if (ret)
 		goto out;
diff --git a/kernel/irq/pipeline.c b/kernel/irq/pipeline.c
index dfa8df1338..6fcf9a04cf 100644
--- a/kernel/irq/pipeline.c
+++ b/kernel/irq/pipeline.c
@@ -3,16 +3,17 @@
  *
  * Copyright (C) 2016 Philippe Gerum  <rpm@xenomai.org>.
  */
-#include <linux/kernel.h>
-#include <linux/sched.h>
+#include <linux/debug_locks.h>
+#include <linux/dovetail.h>
 #include <linux/interrupt.h>
 #include <linux/irq.h>
-#include <linux/irqdomain.h>
 #include <linux/irq_pipeline.h>
 #include <linux/irq_work.h>
+#include <linux/irqdomain.h>
 #include <linux/jhash.h>
-#include <linux/debug_locks.h>
-#include <linux/dovetail.h>
+#include <linux/kernel.h>
+#include <linux/sched.h>
+#include <linux/smp.h>
 #include <dovetail/irq.h>
 #include <trace/events/irq.h>
 #include "internals.h"
@@ -79,6 +80,16 @@ DEFINE_PER_CPU(struct irq_pipeline_data, irq_pipeline) = {
 	},
 };
 
+struct pipeline_percpu_data { };
+static DEFINE_PER_CPU(struct pipeline_percpu_data, pipeline_percpu_data);
+
+static irqreturn_t smp_call_function_ipi_handler(int irq, void *dev_id)
+{
+	smp_flush_oob_call_function_queue();
+
+	return IRQ_HANDLED;
+}
+
 #else /* !CONFIG_SMP */
 
 static struct irq_event_map inband_irq_map;
@@ -1488,6 +1499,30 @@ int notrace run_oob_call(int (*fn)(void *arg), void *arg)
 }
 EXPORT_SYMBOL_GPL(run_oob_call);
 
+struct up_oob_call_tramp {
+	smp_call_func_t func;
+	void *info;
+};
+
+static int __up_oob_call(void *arg)
+{
+	struct up_oob_call_tramp *tramp = arg;
+
+	tramp->func(tramp->info);
+
+	return 0;
+}
+
+/*
+ * up_oob_call_tramp - run_oob_call() trampoline to smp function calls
+ * for uniprocessor configuration.
+ */
+int up_oob_call(smp_call_func_t func, void *info)
+{
+	struct up_oob_call_tramp tramp = { .func = func, .info = info };
+	return run_oob_call(__up_oob_call, &tramp);
+}
+
 int enable_oob_stage(const char *name)
 {
 	struct irq_event_map *map;
@@ -1791,6 +1826,19 @@ void __init irq_pipeline_init(void)
 	 */
 	arch_irq_pipeline_init();
 
+#ifdef CONFIG_SMP
+	/*
+	 * Now that the arch-specific code has prepared for handling
+	 * the oob call function IPI, hook the latter to the generic
+	 * call queue flushing routine.
+	 */
+	if (__request_percpu_irq(CALL_FUNCTION_OOB_IPI,
+				 smp_call_function_ipi_handler,
+				 IRQF_OOB, "Out-of-band function call IPI",
+				 &pipeline_percpu_data))
+		WARN_ON(1);
+#endif
+
 	irq_pipeline_active = true;
 
 	pr_info("IRQ pipeline enabled\n");
diff --git a/kernel/smp.c b/kernel/smp.c
index 0acd433afa..49cd8eabd1 100644
--- a/kernel/smp.c
+++ b/kernel/smp.c
@@ -15,6 +15,7 @@
 #include <linux/percpu.h>
 #include <linux/init.h>
 #include <linux/interrupt.h>
+#include <linux/irq_pipeline.h>
 #include <linux/gfp.h>
 #include <linux/smp.h>
 #include <linux/cpu.h>
@@ -98,6 +99,8 @@ static DEFINE_PER_CPU_SHARED_ALIGNED(struct llist_head, call_single_queue);
 
 static void __flush_smp_call_function_queue(bool warn_cpu_offline);
 
+static void call_function_oob_init(void);
+
 int smpcfd_prepare_cpu(unsigned int cpu)
 {
 	struct call_function_data *cfd = &per_cpu(cfd_data, cpu);
@@ -153,6 +156,8 @@ void __init call_function_init(void)
 	for_each_possible_cpu(i)
 		init_llist_head(&per_cpu(call_single_queue, i));
 
+	call_function_oob_init();
+
 	smpcfd_prepare_cpu(smp_processor_id());
 }
 
@@ -1267,3 +1272,140 @@ int smp_call_on_cpu(unsigned int cpu, int (*func)(void *), void *par, bool phys)
 	return sscs.ret;
 }
 EXPORT_SYMBOL_GPL(smp_call_on_cpu);
+
+#ifdef CONFIG_IRQ_PIPELINE
+
+static DEFINE_PER_CPU_SHARED_ALIGNED(struct llist_head, call_oob_queue);
+static DEFINE_PER_CPU_SHARED_ALIGNED(call_single_data_t, csd_oob_data);
+
+static void __init call_function_oob_init(void)
+{
+	int n;
+
+	for_each_possible_cpu(n)
+		init_llist_head(&per_cpu(call_oob_queue, n));
+}
+
+/*
+ * Runs the oob callbacks queued by remote CPUs. Hard irqs off, oob
+ * stage on entry. The implementation was shamelessly lifted from
+ * __flush_smp_call_function_queue().
+ */
+void smp_flush_oob_call_function_queue(void)
+{
+	call_single_data_t *csd, *csd_next;
+	struct llist_node *entry, *prev;
+	struct llist_head *head;
+
+	WARN_ON(irq_pipeline_debug() && (!hard_irqs_disabled() || running_inband()));
+
+	head = raw_cpu_ptr(&call_oob_queue);
+	if (llist_empty(head))
+		return;
+
+	entry = llist_del_all(head);
+
+	/*
+	 * Run all SYNC callbacks first since senders are waiting for
+	 * us. Callbacks are pushed at head, so walk the list in
+	 * reverse order to preserve the enqueuing sequence.
+	 */
+	entry = llist_reverse_order(entry);
+	prev = NULL;
+	llist_for_each_entry_safe(csd, csd_next, entry, node.llist) {
+		if (CSD_TYPE(csd) == CSD_TYPE_SYNC) {
+			smp_call_func_t func = csd->func;
+			void *info = csd->info;
+
+			if (prev)
+				prev->next = &csd_next->node.llist;
+			else
+				entry = &csd_next->node.llist;
+
+			func(info);
+			csd_unlock(csd);
+		} else {
+			prev = &csd->node.llist;
+		}
+	}
+
+	if (!entry)
+		return;
+
+	/* Finish with ASYNC calls. */
+	prev = NULL;
+	llist_for_each_entry_safe(csd, csd_next, entry, node.llist) {
+		if (prev)
+			prev->next = &csd_next->node.llist;
+		else
+			entry = &csd_next->node.llist;
+		if (CSD_TYPE(csd) == CSD_TYPE_ASYNC) {
+			smp_call_func_t func = csd->func;
+			void *info = csd->info;
+
+			csd_unlock(csd);
+			func(info);
+		}
+	}
+}
+
+/*
+ * smp_call_function_oob - Run a function on a specific CPU, on the
+ *                         out-of-band stage.
+ *
+ * @func: The function to run. This must be fast and non-blocking.
+ * @info: An arbitrary pointer to pass to the function.
+ * @wait: If true, wait until function has completed on the target CPU.
+ *
+ * Hard irqs must be enabled on entry.
+ *
+ * Returns 0 on success, else a negative status code.
+ */
+int smp_call_function_oob(int cpu, smp_call_func_t func, void *info, bool wait)
+{
+	call_single_data_t csd_stack = {
+		.node = { .u_flags = CSD_FLAG_LOCK | CSD_TYPE_SYNC },
+	}, *csd;
+	unsigned long flags;
+	int this_cpu;
+
+	/*
+	 * If hard irqs are off, this call might deadlock if two CPUs
+	 * are waiting for each other to unlock the csd. Make sure to
+	 * detect this.
+	 */
+	WARN_ON_ONCE(hard_irqs_disabled());
+
+	if ((unsigned)cpu >= nr_cpu_ids || !cpu_online(cpu))
+		return -ENXIO;
+
+	this_cpu = hard_get_cpu(flags);
+	if (this_cpu == cpu) {
+		up_oob_call(func, info);
+		goto out;
+	}
+
+	csd = &csd_stack;
+	if (!wait) {
+		csd = raw_cpu_ptr(&csd_oob_data);
+		csd_lock(csd);
+	}
+
+	csd->func = func;
+	csd->info = info;
+
+	if (llist_add(&csd->node.llist, &per_cpu(call_oob_queue, cpu)))
+		irq_send_oob_ipi(CALL_FUNCTION_OOB_IPI, cpumask_of(cpu));
+
+	if (wait)
+		csd_lock_wait(csd);
+out:
+	hard_put_cpu(flags);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(smp_call_function_oob);
+
+#else
+static inline void call_function_oob_init(void) { }
+#endif
-- 
2.30.2

